lib/internal/streams/readable.js

```diff
@@ -1057,6 +1058,30 @@ Readable.prototype.wrap = function(stream) {
   return this;
 };
 
+Readable.prototype.collect = async function({
+  limit = Infinity,
+  rejectAtLimit = true
+} = {}) {
+  const bufs = [];
+  let len = 0;
+  const isObjectMode = this._readableState.objectMode;
+  const encoding = isObjectMode ? false : this._readableState.encoding;
+  for await (const chunk of this.iterator({ destroyOnReturn: rejectAtLimit })) {
+    const chunkLen = isObjectMode ? 1 : chunk.length;
+    if (len + chunkLen > limit) {
+      if (rejectAtLimit) {
+        throw new ERR_STREAM_COLLECT_LIMIT_EXCEEDED(len + chunk.length, limit);
+      } else {
+        this.unshift(chunk);
```

Comment:
this seems... bad? it fails to gather the 2 first chars in the Buffer

Reply:
I think this is correct and needed.

Reply:
That actually seems correct. While the limit is a number of bytes, it's enforced at the chunk level. So if we have 2 chunks of length 5 and 4 respectively, then a limit of 2 would result in an empty buffer and a limit of 7 would result in only the first chunk being in the buffer. If there's a desire to have the limit be at the byte level for buffer streams or character level for string streams, that's not a huge change so I can go ahead and do that.

The "with or without the `unshift()`" might seem weird, but I think that might be a side-effect of piping before await the collect? I think it would be the same if you called `iterator()` instead of collect. You'd get nothing out of that iterator.

Reply:
I'm a bit confused here, I thought limit would be "return once you have X number of things", it seems like it is being implemented as "return if you have more than X number of things?". I'm confused on what the use case of `{limit: 2}` is above; when would I use that?